cmd/binlog-collector/collector/collector.go

```diff
@@ -0,0 +1,235 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "new storage manager")
+	}
+	lastSetFileName := "last-binlog-set"
+	// get last binlog set stored on S3
+	lastSet, err := s3.GetObject(lastSetFileName)
+	if err != nil {
+		return nil, errors.Wrap(err, "get last set content")
+	}
+	bufferSize := int64(16000000)
+	if c.BufferSize != 0 {
+		bufferSize = c.BufferSize
+	}
+
+	return &Collector{
+		storage:         s3,
+		lastSet:         string(lastSet),
+		pxcUser:         c.PXCUser,
+		pxcPass:         c.PXCPass,
+		pxcServiceName:  c.PXCServiceName,
+		lastSetFileName: lastSetFileName,
+		bufferSize:      bufferSize,
+	}, nil
+}
+
+func (c *Collector) Run() error {
+	err := c.newDB()
+	if err != nil {
+		return errors.Wrap(err, "new db connection")
+	}
+	defer c.closeDB()
+
+	err = c.CollectBinLogs()
+	if err != nil {
+		return errors.Wrap(err, "collect binlog files:")
+	}
+
+	return nil
+}
+
+func (c *Collector) newDB() error {
+	host, err := c.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+	pxc, err := db.NewPXC(host, c.pxcUser, c.pxcPass)
+	if err != nil {
+		return errors.Wrapf(err, "new manager with host %s", host)
+
+	}
+	c.db = pxc
+
+	return nil
+}
+
+func (c *Collector) closeDB() error {
+	return c.db.Close()
+}
+
+func (c *Collector) getHost() (string, error) {
+	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+c.pxcServiceName)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrap(err, "get output")
+	}
+	nodes := strings.Split(string(out), "node:")
+	sort.Strings(nodes)
+	lastHost := ""
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ".")
+			lastHost = nodeArr[0]
+		}
+	}
+	if len(lastHost) == 0 {
+		return "", errors.New("cant find host")
+	}
+
+	return lastHost + "." + c.pxcServiceName, nil
+}
+
+func (c *Collector) CollectBinLogs() error {
+	// get last uploaded binlog file name
+	binlogName, err := c.db.GetBinLogName(c.lastSet)
+	if err != nil {
+		return errors.Wrap(err, "get binlog name by set")
+	}
+	list, err := c.db.GetBinLogList()
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+
+	upload := false
+	// if there are no uploaded files we going to upload every binlog file
+	if len(binlogName) == 0 {
+		upload = true
+	}
+
+	for _, binlog := range list {
+		if binlog == binlogName { // this check is for uploading starting from needed file
+			upload = true
+		}
+		if upload {
+			err = c.manageBinlog(binlog)
+			if err != nil {
+				return errors.Wrap(err, "manage binlog")
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Collector) read(r io.Reader) <-chan []byte {
+	bytes := make(chan []byte, c.bufferSize/2)
+	go func() {
+		defer close(bytes)
+		scan := bufio.NewScanner(r)
+		for scan.Scan() {
+			bytes <- scan.Bytes()
+		}
+	}()
+
+	return bytes
+}
```

Comment:
it won't work: https://play.golang.org/p/CcnIAzlIEVr
Let's say your target is 1024 (1Kb). Here you create not channel with capacity 1024 bytes, but channel with a capacity of 1024 slices of bytes (and each slice could have any size). The scanner on each would write into the buffer until it `\n` or EOF - so potentially it could be any amount of bytes. And if the reader from the buffer is slow the amount such buffers will grow up to the channel capacity